[Refactor/Performance] 채팅 시스템 전반적 개선 및 성능 최적화

yunjeooong opened the pull request:
# [Refactor/Performance] 채팅 시스템 전반적 개선 및 성능 최적화

## Description

이 PR은 채팅 시스템 전반에 걸친 다양한 개선사항을 포함하고 있습니다. 주요 목표는 다음과 같습니다:
- **기능 추가**: 프론트엔드 요구 사항 충족 (채팅 목록 조회 시 각 채팅방의 안읽은 메시지 개수와 구매자의 프로필 정보 함께 반환, 채팅 이력은 시간 순서대로 반환)
- **성능 최적화**: 채팅방 목록 조회 및 메시지 로딩 시 발생하는 성능 이슈 해결  
- **코드 품질 개선**: DDD 원칙 적용, 도메인 모델 강화, 명명 규칙 통일  
- **사용자 경험 향상**: 프로필 이미지, 안읽은 메시지 카운트 등 추가 정보 제공  
- **데이터 구조 개선**: MongoDB와 JPA 간의 연계 최적화, 컬렉션 통합  

이러한 개선을 통해 채팅 기능의 성능과 안정성을 크게 향상시키고, 유지보수성을 높였습니다.

---

## Changes

### 1. 성능 최적화
- **N+1 문제 해결**
  - QueryDSL을 활용한 최적화된 쿼리 구현
  - 채팅방 ID 선조회 후 관련 데이터 일괄 로딩 방식 도입
  - `findChatRoomsByUserIdAndRole` 및 `findChatRoomsWithParticipantsAndUsers` 메서드의 2단계 쿼리 패턴 적용

- **MongoDB 최적화**
  - `chat_messages` 컬렉션을 `chat_meta_info`로 통합하여 조회 성능 개선
  - `getChatRoomData` 통합 메서드로 MongoDB 쿼리 횟수 3회 → 1회로 감소
  - 불필요한 데이터 처리 로직 제거

---

### 2. 아키텍처 및 패턴 개선

- **CQRS 패턴 도입**
  - 기존 단일 `ChatRoomService`에 집중된 책임을 분리
  - `ChatRoomCommandService`와 `ChatRoomQueryService`로 역할 분리
  - 각 서비스에 최적화된 쿼리 및 트랜잭션 경계 설정
  - 단일 책임 원칙(SRP) 준수

- **서비스 계층 책임 분산**
  - 비즈니스 로직을 도메인 엔티티로 이동
  - 도메인 이벤트를 통한 간접 의존성 관리 → 결합도 감소
  - 복잡한 로직은 Facade 계층으로 분리하여 협력 구조 개선
  - 도메인별 서비스 구성 (`ChatRoom`, `ChatMessage`, `User` 등)
  - 횡단 관심사 분리를 통한 코드 중복 제거

- **DDD 원칙 적용**
  - 풍부한 도메인 모델 구현 및 애그리게이트 경계 명확화
  - 엔티티와 값 객체의 분리로 불변성 보장
  - 도메인 서비스와 애플리케이션 서비스 역할 구분
  - Repository 패턴 일관 적용

- **코드 가독성 개선**
  - 빌더 패턴 적용 → 14개 파라미터 DTO 생성 단순화
  - 직관적 메서드 네이밍 (`get...` → `find...`)
  - 인터페이스와 구현체 명확 분리 → 테스트 용이성 향상

---

### 3. 새로운 기능 및 화면 개선

- **프로필 정보 향상**
  - 채팅방 조회 시 구매자/판매자 프로필 이미지 URL 제공
  - 상품 이미지 URL 조회 로직 중앙화

- **메시지 관련 기능 개선**
  - 채팅 목록 조회 시 안읽은 메시지 개수 표시
  - 총 안읽은 메시지 수 반환 기능 추가
  - 메시지 페이지네이션 로직 개선

- **역할 기반 조회**
  - 구매자/판매자 역할별 목록 조회 API 구현


---

### 4. 데이터 구조 및 명명 규칙 개선

- **관계 모델링 개선**
  - `ChatRoom-User` 관계를 다대다 매핑으로 개선
  - 엔티티 간 연관관계 최적화

- **필드명 및 메서드명 일관성 확보**
  - `updateAt` → `lastMessageAt` : 마지막 메시지 시간 명확화
  - `isConnected` → `isActive` : 참가자 상태 용어 통일
  - `lastDisconnectedAt` → `lastDeactivatedAt` : 활성화 상태 용어 정리
  - `connect/disconnect` → `activate/deactivate` : 메서드명 통일

- **테스트 코드 개선**
  - 테스트 빌더 패턴 통일
  - UUID 활용으로 테스트 ID 고유성 확보

  
  ## Test
  스웨거와  웹소켓 테스트 결과 기존 처럼 정상적으로 나옴을 확인했습니다.

yunjeooong committed:
feat: 채팅 목록 조회 시 각 채팅방의 안읽은 메시지 개수와 구매자의 프로필 정보 함께 반환

yunjeooong committed:
chore(ChatRoomController): 별 채팅방의 안읽은 메시지 개수를 조회하는 별도의 API 삭제

yunjeooong committed:
refactor: 채팅 메시지 페이지네이션 로직 개선

- 채팅 메시지 조회 시 다음 페이지 존재 여부 확인 로직 개선

yunjeooong committed:
refactor: 불필요한 oldestMessageId 로직을 제거

yunjeooong committed:
perf: 채팅방 목록 조회 성능 대폭 개선

- 성능 측정 로깅 추가로 병목 지점 식별 가능
- 활성 채팅방 메타 정보를 직접 조회하는 최적화된 쿼리 추가
- 필요한 필드만 조회하는 최적화된 ChatRoom 조회 메서드 추가
- 중복 키 처리 로직 추가로 예외 방지
- 중복 상품 ID 제거로 불필요한 조회 방지
- 명확한 주석 추가로 코드 가독성 향상

채팅방 목록 조회 시 발생하는 데이터베이스 쿼리를 최적화하고
불필요한 메모리 내 처리를 제거하여 응답 시간을 크게 단축했습니다.

yunjeooong committed:
fix: 채팅방 목록 조회 버그 수정 및 성능 최적화 유지

- 채팅방 메타 정보가 없을 때 null을 반환하도록 원래 로직 복원
- 연결된 채팅방 필터링 로직 복원
- 최적화된 쿼리 메서드는 유지하여 성능 개선 효과 유지

yunjeooong committed:
feat: 상품 이미지 URL 조회 로직 중앙화

yunjeooong committed:
feat: 누락된 총 안읽은 메세지 갯수 반환 controller 추가

yunjeooong committed:
fix: JangterImages 엔티티 관련 JPQL 쿼리 오류 수정

- 엔티티 이름을 'JangterImage'에서 실제 엔티티명인 'JangterImages'로 수정
- 존재하지 않는 'orderIndex' 속성을 사용한 정렬 조건을 'id'로 변경

yunjeooong committed:
feat: 채팅방 상세 조회 시 구매자/판매자 프로필 이미지 URL 추가

- ChatRoomResponseDTO에 buyerProfileImageUrl, sellerProfileImageUrl 필드 추가
- findChatRoom 메서드에서 구매자와 판매자의 프로필 이미지 URL을 조회하여 반환하도록 수정
- 기존 팩토리 메서드를 유지하면서 프로필 이미지를 포함한 새로운 오버로드 메서드 추가

yunjeooong committed:
refactor(chat): DDD 원칙에 따른 도메인 모델 강화 및 비즈니스 로직 이동

yunjeooong committed:
refactor: 도메인 주도 설계(DDD) 원칙에 맞게 ChatRoomMapper 개선

yunjeooong committed:
feat: 채팅방 목록 조회 N+1 문제 해결을 위한 QueryDSL 구현

- MongoDB와 JPA를 함께 사용하는 구조에서 발생하는 N+1 문제 해결
- ChatRoomRepositoryImpl에 최적화된 쿼리 메서드 구현
- 사용자별 채팅방 조회 시 여러 연관 데이터(참여자, 메시지 등)를 효율적으로 조회하는 로직 추가
- 여러 번의 개별 쿼리 대신 배치 조회 방식으로 성능 개선
- 서비스 레이어에서 각 데이터소스의 결과를 조합하여 완전한 DTO 생성하는 패턴 도입

yunjeooong committed:
refactor: 채팅방 DTO에 빌더 패턴 적용으로 코드 가독성 개선

- ChatRoomDetailDTO 클래스에 롬복 @Builder 애노테이션 추가
- 14개 파라미터를 가진 생성자 대신 빌더 패턴으로 객체 생성 방식 변경
- 필요한 필드만 명시적으로 설정할 수 있도록 개선
- null 값 처리를 위한 @Builder.Default 적용
- 레포지토리와 서비스 클래스의 DTO 생성 로직 수정
- 주석 개선 및 명확한 단계별 설명 추가

yunjeooong committed:
fix: MongoDB MappingException 및 DTO Builder 패턴 수정

yunjeooong committed:
fix: MongoDB MappingException 및 DTO Builder 패턴 수정

yunjeooong committed:
refactor: ChatRoomMapper 중복 메서드 제거 및 서비스 레이어 일관성 개선

yunjeooong committed:
chore(chat): 불필요한 메서드 삭제

yunjeooong committed:
feat: ChatRoom-User 관계를 다대다 매핑으로 개선

yunjeooong committed:
refactor: chat_messages 컬렉션을 chat_meta_info로 통합

yunjeooong committed:
chore(chat): 마이그레이션 코드 삭제

yunjeooong committed:
feat: 온오프라인 로직 구현

yunjeooong committed:
feat: 구매자 판매자 역할에 따른 목록 조회 api 구현

yunjeooong committed:
perf(chatroom): 채팅방 조회 기능 성능 개선 및 안정성 향상

- 역할별 채팅방 목록 조회 최적화
  - findChatRoomsByUserIdAndRole 메서드 2단계 쿼리 구조로 개선
  - 채팅방 ID 선조회 후 참여자 정보 일괄 로드
  - leftJoin 적용으로 연관 데이터 누락 방지

- DTO 변환 로직 안정성 개선
  - 구매자/판매자 정보 누락 시 null 안전성 보완
  - UNKNOWN_USER 상수 활용으로 사용자 정보 처리 일관성 확보

- 모니터링 및 디버깅 기능 강화
  - 상세 로그 추가로 쿼리 성능 및 결과 모니터링 용이

- 기타 성능 개선 사항
  - 채팅방 조회 N+1 문제 해결
  - MongoDB 메타정보 조회 최적화
  - 총 안읽은 메시지 수 계산 성능 향상

yunjeooong committed:
chore: participantUtils 삭제

yunjeooong committed:
perf(chatroom): 채팅방 조회 및 MongoDB 처리 성능 최적화

- 쿼리 최적화
  - 두 단계 쿼리 패턴 도입으로 fetch join 관련 SemanticException 해결
  - findChatRoomsByUserIdAndRole 및 findChatRoomsWithParticipantsAndUsers 메서드 리팩토링
  - 채팅방 ID 우선 조회 → 관련 데이터 일괄 로딩 방식으로 개선

- MongoDB 관련 최적화
  - getUnreadCountMap 메서드 개선으로 불필요한 데이터 처리 감소
  - getChatRoomData 통합 메서드 추가로 MongoDB 쿼리 횟수 3회 → 1회로 감소

yunjeooong committed:
chore: 불필요한 로그 정리

yunjeooong committed:
feat: 채팅 서비스에 CQRS 패턴 도입

yunjeooong committed:
refactor(repository): 'get...' -> 'find...'  네이밍 변경

yunjeooong committed:
fix: Hibernate SemanticException 오류 해결을 위한 Repository 쿼리 최적화

yunjeooong committed:
chore: 불필요한 주석 정리

yunjeooong committed:
test(chatroom): 채팅방 테스트 코드 수정 및 개선

1. ChatRoom.testBuilder() 메서드 대신 표준 builder() 메서드 사용
   - 리팩토링 과정에서 제거된 testBuilder() 메서드 참조 수정

2. 테스트 채팅방 생성 시 고유한 wsRoomId 사용
   - UUID를 활용하여 매 테스트마다 고유한 wsRoomId 생성
   - 'Duplicate entry for key chat_room.UK...' 에러 해결

yunjeooong committed:
refactor(chat): 필드명 및 메서드명 일관성 개선

1. 필드명 변경:
- updateAt → lastMessageAt: 채팅방의 마지막 메시지 시간을 명확하게 표현
- isConnected → isActive: 참가자 상태 용어를 통일하여 일관성 확보
- lastDisconnectedAt → lastDeactivatedAt: 활성화 상태 관련 용어 일관성 유지

2. 메서드명 변경:
- connect/disconnect → activate/deactivate: 상태 변경 메서드 이름 일관성 확보
- 기타 관련 메서드들의 이름을 새로운 필드명에 맞게 수정

yunjeooong committed:
chore: 주석 제거

yunjeooong committed:
chore(StompEventListenerConfig): 디렉토리 위치 변경

yunjeooong committed:
chore(chat): 불필요한 메서드 및 import 문 정리

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/config/StompEventListener.java:
```diff
@@ -33,17 +49,63 @@ public void connectHandler(SessionConnectedEvent event) {
+
+            sessionToChatRooms.putIfAbsent(sessionId, ConcurrentHashMap.newKeySet());
         } else {
             log.warn("STOMP 세션 연결 이벤트 발생했지만 세션 ID가 null입니다");
```
오류 아닌가용?? 연결 이벤트 발생했는데 session id가 null이면?!

yunjeooong replied:
규현님 리뷰 감사합니다!
맞습니다. STOMP 세션 연결 이벤트가 발생했는데 세션 ID가 null인 상황은 분명히 정상적이지 않은 상황입니다. WebSocket 연결이 성공적으로 이루어졌다면 세션 ID가 존재해야 하는 것이 맞습니다.
현재 코드는 단순히 경고 로그만 남기고 있는데, 이는 문제 상황을 파악하기에 충분하지 않을 수 있습니다. 다음과 같이 개선할 수 있을 것 같습니다:
```java
if (sessionId != null) {
    // 기존 코드 유지
} else {
    // 단순 경고가 아닌 오류로 로깅
    log.error("심각한 오류: STOMP 세션 연결 이벤트 발생했지만 세션 ID가 null입니다");
    
  
}
```
참고로, 위에서 이벤트 리스너에서는 DuckwhoException을 발생시키지 않는 것은 다음의 이유가 있습니다!
1. 이벤트 처리 메커니즘의 특성:
 @ EventListener 메서드는 Spring의 이벤트 처리 시스템의 일부입니다. 이벤트 리스너에서 예외가 발생하면 이벤트 처리 파이프라인이 중단될 수 있으며, 다른 리스너들이 이벤트를 처리하지 못할 수 있습니다.
2. 비동기 처리 문제: 
WebSocket은 비동기 통신을 기반으로 하므로, 예외가 발생해도 클라이언트에 적절히 전달되지 않습니다. GlobalExceptionHandler는 HTTP 요청-응답 사이클을 위해 설계되었기 때문에 WebSocket 이벤트에서 발생한 예외를 처리하지 못합니다.
3. 시스템 안정성: 이벤트 리스너에서 처리되지 않은 예외는 이벤트 루프를 중단시킬 수 있어 전체 WebSocket 서비스의 안정성에 영향을 줄 수 있습니다.

beetrbgus replied:
STOMP 에서 에러 발생할 떄 프론트에서도 감지를 할 수 있는걸로 아는데 혹시 아닌가요?

yunjeooong replied:
규현님 너무 늦게 추가 코멘트 확인해 죄송합니다!
@beetrbgus 

 제가 말씀드린 내용은 @ EventListener 메소드에서의 예외 처리에 관한 것이었습니다.

**현재의 @ EventListener 코드는 세션 관리와 로깅 목적으로만 사용되므로, 예외를 던지기보다는 적절한 로그 레벨(error)로 기록하는 것이 더 적합하다고 판단했습니다.**

현재 코드는 @ EventListener 방식으로 구현된 이벤트 핸들러입니다. 이런 방식에서는 예외가 발생해도 STOMP 클라이언트에 자동으로 전달되지 않습니다.

반면, **StompHandler 클래스(ChannelInterceptor 구현)** 에서 발생하는 예외는 STOMP 프로토콜의 ERROR 프레임으로 변환되어 클라이언트에 전달됩니다. 다만 기존에는 AuthenticationServiceException으로 예외를 관리하여 일관성이 떨어지는 문제가 있었습니다. 이에 모든 AuthenticationServiceException을 애플리케이션 전반에서 사용하는 DuckwhoException으로 변경했습니다.


----
간단히 정리하자면:
1. @ EventListener: Spring 내부 이벤트 시스템의 일부로, 여기서 발생한 예외는 STOMP 클라이언트로 전달되지 않음
2. ChannelInterceptor (StompHandler): STOMP 메시지 처리 파이프라인의 일부로, 여기서 발생한 예외는 STOMP ERROR 프레임으로 변환되어 클라이언트에 전달됨
3. 따라서 두 컴포넌트의 예외 처리 전략이 다름 - EventListener는 로깅만, ChannelInterceptor는 DuckwhoException 발생

혹시 추가적으로 이상하거나 궁금하신점이 있으시면 말씀해주세요🙂

yunjeooong committed:
refactor( StompEventListenerConfig): 로깅레벨 수정

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/facade/ChatRoomFacadeService.java:
```diff
@@ -0,0 +1,87 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ChatRoomFacadeService {
```
네이밍에 퍼사드를 직접 기술한 이유가 있을까요?
Service는 사실 파사드 패턴이든 추상 팩토리 패턴이든 상관 없습니당.

그러면 나중에 패턴이 달라질 때는 클래스의 이름도 달라져야 되기 때문에 그냥 어떤 역할을 하는지 정도만 있으면 될 것 같습니다.

yunjeooong replied:
넵! 확인했슴니다! 리뷰 감사합니다

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/facade/ChatRoomFacadeService.java:
```diff
@@ -0,0 +1,87 @@
+public class ChatRoomFacadeService {
+
+    private final ChatRoomCommandService applicationService;
+    private final ChatRoomQueryService queryService;
```
applicationService와
queryService 네이밍 기준을 알 수 있을까요?

yunjeooong replied:
기존 코드에서 구조 변경으로 인해 혼란스러울 수도 있으니 ChatRoomCommandService / ChatRoomQueryService 역할 정리부터 다시 해보겠습니다!

1. 네이밍 규칙과 역할 구분

ChatRoomCommandService (applicationService)
- 주석에 명시된 대로 **Command 파트를 담당하는 서비스**입니다.
- **데이터 변경을 수반하는 작업(생성, 수정, 삭제)**을 처리합니다.
- `@Transactional` 어노테이션만 사용하며, `readOnly` 속성은 없습니다.

 ChatRoomQueryService (queryService)
- 주석에 명시된 대로 **Query 파트를 담당하는 서비스**입니다.
- **데이터 조회 작업만 수행하며**, 상태 변경은 없습니다.
- `@Transactional(readOnly = true)` 어노테이션을 사용합니다.

2. CQRS 패턴 적용

- 이 구조는 **CQRS(Command Query Responsibility Segregation)** 패턴을 따르고 있습니다.
  - **Command**: 시스템의 상태를 변경하는 작업
  - **Query**: 시스템의 상태를 조회하지만 변경하지 않는 작업
- CQRS는 **책임을 명확히 분리**함으로써, 각 역할에 특화된 구현이 가능하게 합니다.

한줄 요약:
ChatRoomCommandService = applicationService (상태 변경 작업 담당)
ChatRoomQueryService = queryService (데이터 조회 작업 담당)

### 3. 그럼 왜 `applicationService`...?

1. DDD에서는 "Application Service"가 도메인 모델 간의 조정과 트랜잭션 관리를 담당하는 계층으로 사용됩되는 걸루 알고 있습니다. 
Command 패턴은 이런 Application Service의 구현 방식 중 하나이므, 좀 더 넓은 개념인 "applicationService"로 명명했습니다.
2. 그런데 좀 더 명확하게 하자면 commandService와
queryService 네이밍  하는게 더 나을 것 같습니다! 제가 처음에는 chatroomCommandservice가 아닌 chatroomapplicationservice로 네이밍했는데 리팩토링 중 부산물인것 같습니다아

혹시 규현님께서는 어떻게 생각하시는지 궁금합니다!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/query/ChatRoomQueryService.java:
```diff
@@ -0,0 +1,282 @@
+    private final ChatMessageQueryService chatMessageQueryService;
+
+
+    private Map<String, Object> getChatRoomData(List<Long> chatRoomIds, Long userId) {
```
Map은 가독성이 낮아 유지보수가 용이하지 못하게 됩니다. 객체를 이용해주세요...

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/document/ChatRoomMetaInfo.java:
```diff
@@ -0,0 +1,239 @@
+    @Builder
+    public ChatRoomMetaInfo(@Param("chatRoomId") Long chatRoomId) {
+        this.chatRoomId = chatRoomId;
+        this.participants = new Participants();
```
객체를 생성할 때 초기화를 해주는게 어떨까요 ??
빈 객체 생성은 객체의 불변성을 깨뜨리기 때문에 좋지 않은걸로 알고있습니당

yunjeooong replied:
규현님의 의견에 전적으로 동의합니다!!  현재 ChatRoomMetaInfo 클래스의 객체 생성과 초기화가 분리되어 있는 방식은 다음과 같은 문제가 있습니다:
- 객체 생성 후 **반드시 `initializeParticipants()`를 호출해야만** 유효한 상태가 되고.
- 이 호출을 누락하면 **참가자 정보가 없는 불완전한 채팅방**이 생성될 수 있는 위험 부담이 있습니다..
- 객체를 사용하는 쪽에서 **초기화 책임을 함께 지게 되어 캡슐화가 약화**됩니다.
-  초기화 순서를 인지하고 따라야 하는 부담이 생겨버립니다.

이를 개선하기 위해 **정적 팩토리 메서드를 사용하는 방식**으로 리팩토링할 예정입니다!:

```java
@Builder
public class ChatRoomMetaInfo {
    private Long chatRoomId;
    private Participants participants;
    private Instant lastMessageAt;

    public static ChatRoomMetaInfo createWithParticipants(
            Long chatRoomId, Long buyerId, Long sellerId) {
        Participants participants = new Participants();
        participants.addParticipant(buyerId, JangterChatRole.BUYER);
        participants.addParticipant(sellerId, JangterChatRole.SELLER);

        return ChatRoomMetaInfo.builder()
                .chatRoomId(chatRoomId)
                .participants(participants)
                .build();
    }
}
```

이렇게 하면 
- 객체 생성 시점에 모든 필수 정보가 초기화되어 **완전한 상태를 보장**할 수 있고
- 객체 사용자 입장에서 **한 번의 호출로 안전하게 초기화된 객체를 사용할 수 있습니다.** 

좋은 리뷰 다시한번 감사합니다☺️

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/document/ChatRoomMetaInfo.java:
```diff
@@ -0,0 +1,239 @@
+    /**
+     * 특정 사용자의 안 읽은 메시지 개수를 초기화합니다.
+     */
+    public boolean resetUnreadCount(Long userId) {
```
readAllMessages 에서 사용되는 값들이 전혀 사용되지 않는데 왜 boolean 값을 반환하나요?!?

yunjeooong replied:
헉! 놓친 부분 찾아주셔서 감사합니다... resetUnreadCount 메서드의 boolean 반환값을 유지하고, readAllMessages 메서드에서 이 값을 활용하여 실패 상황을 명시적으로 처리하도록 수정했습니다.

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/document/ChatRoomMetaInfo.java:
```diff
@@ -0,0 +1,239 @@
+    /**
+     * 여러 채팅방에 대한 특정 사용자의 총 안 읽은 메시지 개수를 계산합니다.
+     */
+    public static int calculateTotalUnreadCount(List<ChatRoomMetaInfo> chatRooms, Long userId) {
```
reduce가 아닌 count를 세면 되지 않을까 싶습니당. 
그리고 totalUnreadCound를 계산하는게 아니라 세는거라 getTotalUnreadCount로 써도 되지 않을까요!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/document/ChatRoomMetaInfo.java:
```diff
@@ -0,0 +1,239 @@
+     */
+    public Optional<ChatMessage> findMessageById(String messageId) {
+        return messages.stream()
+                .filter(msg -> msg.getId().equals(messageId))
```
약어는 유지보수에 좋지는 않습니당. 물론 msg가 message다라고 유추는 되지만 
합의된 약어가 아니라면 좋지는 않습니당!
 그냥 message 로 풀어도 가독성을 해치지 않을 것 같습니다!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/document/ChatRoomMetaInfo.java:
```diff
@@ -0,0 +1,239 @@
+     * @param limit 조회할 메시지 수
+     * @return 조건에 맞는 메시지 목록
+     */
+    public List<ChatMessage> getMessagesBeforeTime(LocalDateTime before, int limit) {
```
이거는 어떨 때 사용하나요?
for문의 용도를 잘 모르겠습니다 왜 count < limit으로 하는지 이해가 잘 안돼요!

yunjeooong replied:
규현님 리뷰주셔서 감사합니다:)

이 메서드는 채팅 메시지 무한 스크롤 기능 중 하나 입니다! 사용자가 채팅 내역을 위로 스크롤할 때 이전 메시지를 추가로 불러오는 역할을 합니다!

for 문의 동작 방식과 count < limit 조건의 의미는 다음과 같습니다:
**1. 역순 탐색**
for (int i = messages.size() - 1; ...)는 메시지 리스트의 끝에서 시작해 앞쪽으로 거꾸로 탐색합니다. 채팅 앱에서는 일반적으로 최신 메시지가 아래쪽  오래된 메시지가 위쪽에 위치하기 때문에 이런 탐색 방향으로 구현했습니다!
**2. 필터링 조건**
message.getSentAt().isBefore(before)는 지정된 시간보다 이전에 보낸 메시지만 필터링하는 조건입니다. 이는 커서 기반 페이징에서 기준점 역할을 하며 “이 시간 이전의 메시지”를 찾아주는 로직입니다.
(예를 들어, 유저가 현재 '2025-11-01 10:30:00'에 전송된 메시지를 화면에서 가장 위에 보고 있다면 그보다 이전(10:29:59 이전)에 전송된 메시지들만 추가로 가져오는 것입니다. 이렇게 하면 사용자가 이미 본 메시지를 제외하고 그 이전의 메시지만 정확히 가져올 수 있습니다!)

**3. 개수 제한 (count < limit)**
무한 스크롤 방식에서는 한 번에 너무 많은 메시지를 가져오면 성능에 부담이 생길 수 있습니다. 따라서 최대 limit 개까지만 불러오도록 제한합니다. 이 조건이 없다면 모든 이전 메시지를 한 번에 가져오게 되어 효율적이지 않다고 생각했습니다!

이 메서드는 ChatMessageQueryService.getChatMessages()를 통해 호출되고  프론트엔드에서는  스크롤 이벤트를 통해 동작하게 됩니다!☺️

혹시 추가적으로 궁금하시거나 이상한 부분있으면 리뷰 부탁드리겠습니다!

beetrbgus replied:
그래서 궁금한건
해당 채팅방을 MongoDB에서 조회할!!때!!! Query로 
해당 메시지 주고받은 이전 시간의 메시지를 조회하는게 아닌
전체 메시지를 조회해서 거기서 짤라서 붙이는 형식같아 보여서 
왜 그렇게 하는건가 궁금했습니다.

yunjeooong replied:
억!!.. 저 바본가봐요
리뷰 감사합니다! 지적하신 부분이 정확합니다. 현재 구현은 모든 메시지를 메모리로 가져온 후 자바 코드에서 필터링하는 방식인데 데이터가 많아질수록 아주아주 비효율적입니다...

MongoDB의 쿼리 기능을 활용하여 다음과 같이 개선하겠습니다!
1. MongoDB 쿼리로 직접 특정 시간 이전 메시지만 필터링
2. 정렬과 limit을 데이터베이스 레벨에서 처리
3. 필요한 데이터만 애플리케이션으로 전송

또한 적절한 인덱스를 추가하여 쿼리 성능도 최적화하겠습니다. 다시 한번 리뷰 감사합니다!

yunjeooong replied:
[commit 51](https://github.com/taku-project/taku-backend/pull/130/commits/7bcb16f463e290ae3687b6333b19158d5251780d)
수정했습니다!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/query/ChatMessageQueryService.java:
```diff
@@ -0,0 +1,126 @@
+                LocalDateTime sentAt = targetMessage.get().getSentAt();
+                messages = metaInfo.getMessagesBeforeTime(sentAt, limit);
+            } else {
+                messages = new ArrayList<>();
```
else를 사용하지 말고 위의 List<ChatMessage> messages 에서 선언과 초기화를 동시에 하면 어떨까요?

yunjeooong replied:
넵!  List messages 에서 선언과 초기화를 동시에 하니 중첩 구조가 한 단계 줄어들어드네요! 대체 과거의 저는 왜 저렇게 했을까요,.. 감사합니다!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/query/ChatRoomQueryService.java:
```diff
@@ -0,0 +1,282 @@
+            unreadCountMap.put(chatRoomId, unreadCount);
+        }
+
+        for (Long chatRoomId : chatRoomIds) {
```
이건 또 왜 별도로 있는건가요?? 위의 for문에서 가능하지 않나요?!

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/query/ChatRoomQueryService.java:
```diff
@@ -0,0 +1,282 @@
+        }
+
+
+        Map<Long, User> userMap = buildUserMap(chatRoom);
```
Map을 사용하는 이유가 있을까요?? 
ap 을 사용했을 때 이점이 뭔지 한 번 말씀해주시겠어용?

yunjeooong replied:
1. Map은 O(1) 시간 복잡도로 조회가 가능하므로  많은 채팅방을 처리할 때 매우 효율적이라고 생각했어요.
 각 Map은 관련 ID(주로 채팅방 ID나 사용자 ID)로 특정 유형의 데이터를 인덱싱하여 관련 정보에 빠르게 접근할 수 있게 구현했습니다.
2. 그리고 데이터베이스 쿼리 감소를 할수 있게 해서 효율성을 높이려 했아요.
필요한 모든 데이터를 미리 수집하고 Map으로 구성함으로써 데이터베이스 쿼리 횟수를 최소화합니다. 각 채팅방마다 개별적으로 데이터베이스를 조회하는 대신 한 번에 데이터를 가져와서 빠른 접근을 위해 구조화했습니다.

yunjeooong replied:
지금 생각해보니 dto로 빼서 관리하는게 나을 수도 있을것 같습니ㅏㄷ.

beetrbgus replied:
DB 쿼리 횟수를 최소화 한다는게 뭘까요?? 
buildUserMap 부분의 메서드의 내용과 위의 설명은 잘 이해가 안가요

yunjeooong replied:
아 이부분은 buildusermap 부분에 대한 게 아니라 카톡에서 말슴드렸던 내용인데 여기에 달았네요. 
buildChatRoomResponseDTOs 이메서드는 map을 왜 이렇게 많이 썼냐에 대한 답변이었습니다. 😅

yunjeooong replied:
네, Map을 사용한 이유는. 사용자 ID로 바로 정보를 찾을 수 있어서 빠르고, 같은 사용자가 여러 채팅방에 있을 때도 한 번만 저장하면 되니까 효율적이라고 생각했습니다. 특히 채팅방 목록을 불러올 때 여러 사용자 정보를 매번 DB에서 가져오는 즉, 디비 부하를 피하려는 목적이 제일 컸어요
지금은 UserInfoCollectionDTO라는 dto를 만들어서 개선했습니다! 이 방식으로 사용자 정보를 깔끔하게 DTO로 감싸서 필요한 정보만 주고받을 수 있고 내부 데이터가 함부로 변경되지 않도록 하였습니다. 가독성도 좋아진 것 같습니다!
항상 리뷰 감사합니다 규현님:)

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/domain/mapper/ChatRoomMapper.java:
```diff
@@ -0,0 +1,108 @@
+            return null;
+        }
+
+        Long buyerId = buyer != null ? buyer.getUserId() : null;
```
여기서 개별적으로 하신 이유가 있을까요?
ChatRoomResponseDTO에서 
buyer : {
  id : "",
  nickname : "" 
} 

이렇게 하면 코드 상으로 buyer != null 와 셀러 
이부분을 하나로 퉁칠 수 있을 것 같고,
프론트에서도 확인하기 편할 것 같은데 객체 정보가 아닌 나열식이라서 코드도 줄어들 수 있을 것 같습니다.

yunjeooong replied:
말씀하신 대로 UserInfoDTO라는 dto를 별도로 추출하여 개선했습니다!
사용자 정보를 객체로 표현해 코드의 가독성을 향상시키고 buyer != null과 같은 검사를 UserInfoDTO.from(buyer)로 간소화했습니다. 추가로 UserInfoCollectionDTO도 구현하여 여러 사용자 정보를 효율적으로 조회하는 N+1 쿼리 문제도 해결했습니다~
다시한번 리뷰주셔서 감사합니다:)

yunjeooong committed:
chore(chatservice): 네이밍 변경 및 facade 디렉토리 삭제

yunjeooong committed:
refactor(facade): applicationservice-> commandservice 네이밍 변경

yunjeooong committed:
refactor(chatroom): ChatRoomMetaInfo 객체 생성 로직 개선

- ChatRoomMetaInfo 생성자에 participants 매개변수 추가
- createWithParticipants 정적 팩토리 메서드 도입으로 객체 생성과 초기화를 한 번에 처리
- 기존 initializeParticipants는 호환성 유지를 위해 일시적으로 유지
- ChatRoomCommandService의 createAndSaveChatRoomMetaInfo 메서드에서 builder 방식 제거, 팩토리 메서드로 변경

yunjeooong commented on src/main/java/com/ani/taku_backend/chatroom/service/query/ChatRoomQueryService.java:
```diff
@@ -0,0 +1,282 @@
+                        articleImageMap))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
```
> 1. Map은 O(1) 시간 복잡도로 조회가 가능하므로  많은 채팅방을 처리할 때 매우 효율적이라고 생각했어요.
>    각 Map은 관련 ID(주로 채팅방 ID나 사용자 ID)로 특정 유형의 데이터를 인덱싱하여 관련 정보에 빠르게 접근할 수 있게 구현했습니다.
> 2. 그리고 데이터베이스 쿼리 감소를 할수 있게 해서 효율성을 높이려 했아요.
>    필요한 모든 데이터를 미리 수집하고 Map으로 구성함으로써 데이터베이스 쿼리 횟수를 최소화합니다. 각 채팅방마다 개별적으로 데이터베이스를 조회하는 대신 한 번에 데이터를 가져와서 빠른 접근을 위해 구조화했습니다.
[#r1999822216](https://github.com/taku-project/taku-backend/pull/130/files#r1999822216)

위에서 제가 위치를 헷갈려 다른곳에 다가 답변을 달았네요..ㅎㅎ

기존에는
채팅방 목록 조회(1번) + 각 채팅방마다 마지막 메시지 개별 조회(N번) = N+1번 쿼리
로 되어 있어 N+1 번의 db 쿼리가 발생했습니다. 
여러 엔티티 간의 관계 데이터를 효율적으로 조회하기 위해서 기존에 Map을 사용했는데요, 
예를 들어, 채팅방 10개와 각 채팅방의 메타정보, 사용자 정보 등을 조회할 때, 목록에서 반복 순회(O(n))하는 대신 Map으로 즉시 조회(O(1))가 가능합니다.

그런데 규현님께서 카톡에서 말씀해주신것 처럼 가독성이 떨어지고 불변성이 떨어지는 등 여러 문제를 야기하므로 되므로 
 Map 대신 채팅방 데이터를 담는 DTO만들어  개선할 예정입니다. 이렇게 해서 일괄 조회 방식과 효율적인 데이터 접근 구조는 유지하고 가독성과 타입 안전성은 높이겠습니다. 

추가적으로, 현재 확인해본 결과 채팅방 목록에 페이지네이션이나 무한 스크롤 방식이 적용되지 않고 한번에 데이터를 가져오므로 프론트엔드분과 협의해 둘 중 하나의 방식을 선택해 적용하겠습니다!   
👉🏻무한 스크롤 방식으로 결정했습니다!

@beetrbgus

yunjeooong committed:
refactor: Map 대신 Value Object 사용하도록 전체 코드 리팩토링

yunjeooong committed:
chore(chat): 불필요한 메서드와 import문 제서

yunjeooong committed:
refactor: 객체지향 원칙에 따른 채팅 관련 클래스 리팩토링

- VO에 팩토리 메서드 추가하여 객체 생성 책임 강화
  - ChatRoomMessages.fromMetaInfos
  - UnreadMessageCounts.fromMetaInfos

- 메서드명 및 클래스명 개선
  - prepareRelatedData → aggregateChatRoomData
  - ChatRoomDataBundle → ChatRoomCompositeDTO

- 중복 코드 제거 및 책임 명확화
  - 채팅방 접근 권한 검증 로직 ChatRoomQueryService로 일원화
  - ChatMessageQueryService는 ChatRoomQueryService에 위임

- 데이터 관리 단일화
  - 상태 관련 로직은 MongoDB 문서(ChatRoomMetaInfo)에서만 관리
  - JPA 엔티티에서 중복 상태 관리 로직 제거

yunjeooong committed:
chore: 불필요한 메서드 제거

yunjeooong committed:
refactor: ArticleImage VO를 Map 대신 ArticleImageItem을 사용하도록 변경

yunjeooong committed:
refactor: Value Object 클래스들을 Map에서 List 기반 객체 모델로 변경

yunjeooong committed:
refactor(chat): 문자열 상수화

yunjeooong committed:
fix: 채팅방 목록 조회 시 발생하는 NullPointerException 수정

yunjeooong committed:
var타입을 명시적 타입 선언(User)으로 변경

yunjeooong committed:
chore(chatroom): dto를 domain layer에서 분리

yunjeooong committed:
feat: MongoDB 쿼리 최적화로 채팅 메시지 로딩 성능 개선

- ChatRoomMetaRepository에 시간 기반 메시지 필터링 쿼리 메소드 추가
- ChatMessageQueryService에 최적화된 getMessagesBeforeTime 구현
- 기존 메모리 내 필터링 방식에서 MongoDB 쿼리 기반 필터링으로 변경

yunjeooong committed:
refactor(StompHandler): AuthenticationServiceException을 DuckwhoException으로 변경

yunjeooong committed:
fix:충돌 해결

yunjeooong committed:
fix: ChatRoomController에서 @PathVariable에 명시적 파라미터 이름 추가

yunjeooong committed:
refactor: 채팅 관련 API 페이징 방식 개선

- 채팅방 목록 API: Spring Data의 Slice 인터페이스를 사용한 표준 페이징 구현
  - page/size 파라미터 지원
  - 불필요한 커스텀 페이징 메서드 제거
  - 정렬 기준 명확화 (업데이트 시간 내림차순)

- 채팅 이력 API: messageId 기반 커서 페이징 유지
  - 메시지 특성에 적합한 커서 기반 페이징 방식 적용
  - hasMore 플래그를 통한 무한 스크롤 지원

yunjeooong committed:
fix: 파라미터 이름 누락으로 인한 API 오류 해결

yunjeooong committed:
chore: 불필요한 메서드 삭제

yunjeooong committed:
Merge branch 'develop' into hotfix/250310_chat
